zebrad/src/components/mempool.rs

```diff
@@ -372,20 +372,36 @@ impl Service<Request> for Mempool {
             // Collect inserted transaction ids.
             let mut send_to_peers_ids = HashSet::<_>::new();
 
+            let best_tip_height = self.latest_chain_tip.best_tip_height();
+
             // Clean up completed download tasks and add to mempool if successful.
             while let Poll::Ready(Some(r)) = tx_downloads.as_mut().poll_next(cx) {
                 match r {
-                    Ok(tx) => {
-                        let insert_result = storage.insert(tx.clone());
-
-                        tracing::trace!(
-                            ?insert_result,
-                            "got Ok(_) transaction verify, tried to store",
-                        );
-
-                        if let Ok(inserted_id) = insert_result {
-                            // Save transaction ids that we will send to peers
-                            send_to_peers_ids.insert(inserted_id);
+                    Ok((tx, expected_tip_height)) => {
+                        // # Correctness:
+                        //
+                        // It's okay to use tip height here instead of the tip hash since
+                        // chain_tip_change.last_tip_change() returns a `TipAction::Reset` when
+                        // the previous block hash doesn't match the `last_change_hash` and the
```

Comment:
Nitpick: it might help to explain the consensus rule, rather than the current implementation of it

```suggestion
                        // chain_tip_change.last_tip_change() returns a `TipAction::Reset` when
                        // the best chain changes (which is the only way to stay at the same height), and the
```